faucet/src/service/mod.rs

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{Faucet, SimpleFaucet};
+use async_trait::async_trait;
+use std::sync::Arc;
+use tracing::debug;
+
+pub use self::faucet_service::{FaucetRequest, FaucetResponse};
+
+mod faucet_service;
+
+#[async_trait]
+pub trait FaucetService {
+    async fn execute(self, faucet: &(impl Faucet + Send + Sync)) -> FaucetResponse;
+}
+
+pub struct Service<F = SimpleFaucet>
+where
+    F: Faucet + Send + Sync,
```

Comment:
@patrickkuo @bmwill I don't have a good understanding of async/concurrent stuffs in Rust yet but plan to read up more this weekend. In the meanwhile, I would really appreciate it if you guys have any concrete feedback on improving/fixing these code involving send/sync/Arc<Mutex>.

Reply:
I'm not sure I understand the need for the extra level of indirection with this Service type + the FaucetService trait.

Can't you just drop the trait and this type and instead just use a single `Faucet` struct that can handle the request?

Reply:
Yes I could have. The idea is for separation of concerns between
- Faucet struct handles transferring/splitting coins
- Service handles rate-limiting(by ip, sui address, etc). This part has not been implemented yet. It can also handle different requests(e.g., flexible amount request, on line 60) using the same Faucet interface.

Reply:
I don't think you need to get this complicated yet. The way tower works is that things are composable. Ie your `post` request is the leaf of the service and you can add layers around that for doing the rate limiting so it seems overkill to wrap your actually logic handling the request in another layer of indirection.